(SERVER-297) Consolidate JRuby environment handling

jeffmccune opened the pull request:
Without this patch the manner in which environment variables are handled in the
JRuby portions of puppet-server vary across the use cases of the `puppetserver
{gem,irb,ruby}` command line utilities in addition to the service itself.  This
is a problem because the behaviors have not kept in sync with one another over
time so there has been divergence at times, e.g. `puppetserver gem env`
currently fails because PATH is unset in that scenario but set in others, e.g.
the production service.

This patch addresses the problem by consolidating all of the environment
handling behavior into a single authoritative method.  The behavior itself
remains unchanged, GEM_HOME is set based on the config file value, PATH and
HOME are the only variables allowed to flow through to JRuby unmodified, all
other variables are filtered out, and JARS_REQUIRE and JARS_NO_REQUIRE are set
appropriately.

In addition to the environment handling, this patch also consolidates the JRuby
load path handling in a similar fashion across the scenarios of command line
utilities and the production service.

jeffmccune commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -43,25 +41,55 @@
+                    (cons ruby-code-dir)
+                    (cons (str jruby-home "/lib/ruby/1.9"))
+                    (cons (str jruby-home "/lib/ruby/shared"))
+                    (cons (str jruby-home "/lib/ruby/1.9/site_ruby")))]
```
@camlow325 One question I had about this...  The ticket says:

> For Server 1.X only, add in the "front-loading" of the load path with embedded JRuby path libs, as is done for the irb subcommand today - see https://github.com/puppetlabs/puppet-server/blob/puppet-server-1.0.8/src/clj/puppetlabs/puppetserver/cli/irb.clj#L20-L24.

However, in the master branch both the `irb` and `ruby` commands, but not `gem` front-load these three directories relative to `JRUBY_HOME`.  I left them in the consolidated implemenetation, but I was wondering if you could clarify the "for 1.X only" piece.

jeffmccune commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -15,21 +15,19 @@
-  This directory lives under src/ruby/"
-  "puppet-server-lib")
+  "The name of the directory containing the ruby code in this project."
+  "src/ruby/puppet-server-lib")
```
@camlow325 or @cprice404 or whomever feels confident about this...  Two questions I had while working with this bit of code:

 1. The path has been relative in the implementation up until this patch.  I'm not sure that's wise and/or secure for production since this is loading library code.  When my patch uses `ruby-code-dir` now, which is the only place it is used, the value is made absolute and canonical.  The question is: is this reasonable?  Is it a pre-mature improvement?
 2. Given that the path is relative and appears to be only intended to run things from source, does it make sense to include this path in the packaged releases of puppetserver?  I'm trying to come up with how this would be used in a production deployment and am falling short.

nwolfe replied:
This is the path into the ruby code from the root of the `puppet-server-release.jar`.  

@jeffmccune I bet if you built packages with lein ezbake with this change that the CLI tools might not work.

jeffmccune replied:
Hrm.  Yeah, I wondered about that...

@nwolfe Putting that (it's broken) issue aside, does it make sense to you to have a relative directory on a LOAD_PATH?

jeffmccune replied:
Well, after some digging I figured out how this works.  It's documented at a high level, but sufficient enough that I feel like I understand it, in the [LoadService](http://jruby.org/apidocs/org/jruby/runtime/load/LoadService.html) JRuby java doc.

Basically, JRuby will search each element of the classpath by taking the name being required, map it onto each element of Ruby's LOAD_PATH, then map it again onto each element of the classpath.  This seems like a ridiculous huge number of permutations, but hey it seems to work!

More concerning to me is this behavior seems to assume there will be relative directories located on the Ruby LOAD_PATH.  That seems terrible and just plain wrong, but it's also how upstream JRuby fundamentally behaves, so we're at least relying on intended behavior upstream, if not sound behavior.

KevinCorcoran replied:
@jeffmccune Just to be clear - this is a relative path **inside the .jar we build**, not a path on the filesystem.  In general, yeah, a relative path on LOAD_PATH seems funky, but in this case it seems correct to me.

jeffmccune replied:
@KevinCorcoran Yeah, understood that it's a relative path inside the jar, but the behavior is that it's added to the $LOAD_PATH (which is funky) and JRuby looks in JAR files by pre-pending each element of the LOAD_PATH to the value passed to "require" then peeking inside the JAR root.

So, even though it's used for a JAR file, it's added to the LOAD_PATH which affects both the behavior of filesystem and JAR resource loading.

KevinCorcoran replied:
ohhh, my.  Gross.  Well, thanks for the explanation.

![](http://i3.kym-cdn.com/entries/icons/original/000/015/976/homer-simpson-bush-gif.gif)

jeffmccune commented on src/clj/puppetlabs/puppetserver/cli/gem.clj:
```diff
@@ -1,18 +1,10 @@
-        "JARS_REQUIRE" "false"))
-    (.runScriptlet "require 'jar-dependencies'")
-    (.runScriptlet "load 'META-INF/jruby.home/bin/gem'")))
+  (jruby-core/cli-run! config "gem" args))
```
I suppose I could kill this function entirely and reduce this to:

```clojure
(defn -main [& args] (cli/run jruby-core/cli-run! "gem" args))
```

Sane?

nwolfe replied:
I wouldn't be opposed to squashing these like you proposed.  Then at least all these minimal CLI namespaces will look the same, where they just call `cli/run` with a `jruby-core/cli-{run!,ruby!}` function.  No big deal though

camlow325 commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -250,3 +279,20 @@
     ;; if we get here, we got a Retry, so we just put it back into the pool.
     (.putFirst (:pool instance) instance)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
I think there are some other functions in this namespace which are called from outside of this namespace, e.g., `create-pool-from-config` appears to be called from `jruby-puppet-agents`.  I like the idea of adding the "public" section here, but I think it would be good to move any of the other functions which effectively are public down into this new section.

jeffmccune replied:
Addressed in be6ed72

camlow325 commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -43,25 +41,56 @@
-                             "JARS_REQUIRE" "false"}
-                            jruby-puppet-env))))
+  (let [jruby-home (.getHomeDirectory scripting-container)]
+    (doto scripting-container
```
The `scripting-container` initialization here and the `jruby-config` initialization in `new-main` below look very similar - other than just a difference in the underlying types being used.  What would you think about consolidating these into a single function which would operate either on the `ScriptingContainer` or `RubyInstanceConfig` passed into it?  If the idea of taking advantage of the duck-typing here seems too distasteful, could we at least factor out a common helper function for `compat-version` that both this function and `new-main` could call into to get the same `CompatVersion`?

cprice404 replied:
Wow, that's a very unfortunate API design on their part :(

If they just had an interface that was implemented by both `ScriptingContainer` and `RubyInstanceConfig`... :(

The duck typing idea bugs me, quite a bit, but it does seem like our tests would catch it if the interface changed on either of the two objects in a future JRuby release.

I'm on the fence as to which I think is better, but I do agree with you that it would be a good idea to wrap the `compat-version` and `compile-mode` values in a function or a var and reference that from both of these initializers.

jeffmccune replied:
fixed in  ef931c9 (maint) Add compile-mode, compat-version defs

camlow325 commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -43,25 +41,56 @@
+    (doto scripting-container
+      (.setLoadPaths (managed-load-path ruby-load-path jruby-home))
+      (.setCompatVersion (CompatVersion/RUBY1_9))
+      (.setCompileMode RubyInstanceConfig$CompileMode/OFF)
```
Any reason we would do this via the `ScriptingContainer` but not via a `RubyInstanceConfig` (through `new-main`).  Seems like we should be doing the same thing in both places - again through a common helper function that both paths would use so that we would just have to change the default value in one place.

jeffmccune replied:
Yes, basically I didn't want to change the way we were initializing a ScriptingContainer which has been to not inject a RubyInstanceConfig into the initializer and instead configure it after initialization.

However, I needed to do so for the use of Main since that was the only way I could figure out to get the jruby home.

Upstream JRuby [says](https://github.com/jruby/jruby/blob/1.7.20/core/src/main/java/org/jruby/Main.java#L178-L184) that Main should only be used for command-line entry points for JRuby, and to use other mechanisms when embedding JRuby into an application.

That's why I kept both Main and ScriptingContainer around.

jeffmccune replied:
I think the best way to solve this if we really want to is to figure out if we could have a single function that produces RubyInstanceConfig instances, then see if we can inject that instance into a ScriptingContainer like we're already doing it for Main.

cprice404 commented on src/clj/puppetlabs/services/jruby/jruby_puppet_schemas.clj:
```diff
@@ -172,3 +181,12 @@
+
+(def JRubyMainStatusOrNil
+  (schema/either JRubyMainStatus
+    (schema/pred nil? 'nil?)))
```
this could just be `schema/maybe JRubyMainStatus`, right?

jeffmccune replied:
`maybe`...  I'm sure you're right, I just started writing schemas for functions so I'll check and fix it up if so.

cprice404 commented on test/unit/puppetlabs/services/jruby/jruby_puppet_core_test.clj:
```diff
@@ -13,6 +14,37 @@
+         (.flush capturing-print-stream#)
+         (swap! return-map# assoc :out (.toString captured-output#))
+         (System/setOut system-output#)))
+     @return-map#))
```
These macros are doing some pretty hard-core stuff (System/setOut, etc.) that makes me nervous.  That said, I looked at the tests and they do seem valuable, so I think I'm good with them... but... yikes.

jeffmccune replied:
Yeah, they make me nervous too.  The original implementation was using functions but was really awkward.  I'd like to use a binding in clojure, but my assumption is that won't help with `System.out` and `System.in`.

My main concerns are around thread safety, but I can't imagine anything else mucking with System.out unless we do something like [parallel-test](https://github.com/aredington/parallel-test) which requires tests to be marked as `^:parallel` before they'll run concurrently.

My thinking was this isn't really any worse than how we're doing other things like TCP 8140, but it's also far from ideal and has lots of room for improvement.

cprice404 commented on test/unit/puppetlabs/services/jruby/jruby_puppet_core_test.clj:
```diff
@@ -13,6 +14,37 @@
+         captured-output# (new ByteArrayOutputStream)
+         capturing-print-stream# (new PrintStream captured-output#)]
+     (try
+       (System/setOut capturing-print-stream#)
```
I think that in Clojure you might normally do this with something like:

`(binding [*out* capturing-print-stream#]`

But, not sure if that would work for JRuby, so if we're messing w/stdin and stdout at all during the tests, it probably doesn't matter much which way we do it.

I wonder if it would be worth putting these tests into the `integration` test suite since they're a little riskier (because they much with stdout/stderr)?  Not sure about that one either.

jeffmccune replied:
I've got them marked as integration tests already and was wondering if that alone was enough to warrant they move into the integration sub-directory.

nwolfe replied:
> move into the integration sub-directory.

FWIW I don't see the value of making the distinction between unit and integration tests at the directory level.
:+1: for keeping the test here

nwolfe replied:
Might be worthwhile to mention in the docstring that redirecting streams like this is necessary for JRuby, considering it's not quite standard Clojure (like Chris was mentioning).

jeffmccune replied:
Fixed in 72c6d4bd9e0d9a8c81d8b0b37139304a646bc805, thanks @nwolfe

cprice404 commented on test/unit/puppetlabs/services/jruby/jruby_puppet_core_test.clj:
```diff
@@ -13,6 +14,37 @@
+  `(let [return-map# (atom {})
+         system-output# (System/out)
+         captured-output# (new ByteArrayOutputStream)
+         capturing-print-stream# (new PrintStream captured-output#)]
```
nitpick: I think the `(ClassName.)` constructor form is more prevelant in Clojure than the `(new ClassName)` form.

jeffmccune replied:
Noted, thanks!  I'll use the conventional form from here on out.

nwolfe replied:
FWIW I like `(new CLASS...)` better as it looks like normal functional Clojure, whereas `(CLASS. ...)` looks like Java interop....which I suppose you could argue is what this code is anyway :smile:

jeffmccune replied:
I wasn't going to argue the point, but I'm with @nwolfe, I found `(new Foo)` easier to read and understand as a newbie than `(Foo.)` which is less obvious about what's happening.

cprice404 commented:
I think I'm +1, except for the questions about possibly DRY'ing up the ScriptingContainer/RubyInstanceConfig stuff a bit.

jeffmccune commented:
> I think I'm +1, except for the questions about possibly DRY'ing up the ScriptingContainer/RubyInstanceConfig stuff a bit.

Cool, at this point the only thing that remains to change beyond addressing comments is to add 1 beaker test for each of the subcommands, which I hope to do today.

I also have a TODO item to build packages off this topic branch and validate we don't run into the hipchat gem problem described in [SERVER-538](https://tickets.puppetlabs.com/browse/SERVER-538).  If we do run into the issue my plan is to just pull out d7574cc.

jeffmccune commented:
Acceptance test added, this should be ready now.  No longer "for review" but still needs the normal review.

Chris, this is largely the same as when you looked at it earlier.

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -15,53 +15,82 @@
+
+(def compat-version
+  "The JRuby compatibility version to use"
+  (CompatVersion/RUBY1_9))
```
Are these extra parens necessary? Just curious, since we don't have them for `compile-mode` above. I can't remember if there's a subtle difference

jeffmccune replied:
Yes, CompatVersion/RUBY1_9 is a method, so we need to call it to get the value.

KevinCorcoran replied:
Hmm, actually, it's not a method, it's an enum value ...
```java
public enum CompatVersion {

    RUBY1_8, RUBY1_9, RUBY2_0, BOTH;
```
... which means that the parens can be removed, I think.  Should be same as accessing a field.

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -71,7 +100,7 @@
    :post [(instance? ScriptingContainer %)]}
   (-> (ScriptingContainer. LocalContextScope/SINGLETHREAD)
-      (prep-scripting-container ruby-load-path gem-home)))
+    (prep-scripting-container ruby-load-path gem-home)))
```
:cry:

jeffmccune replied:
Yeah, sorry.  This snuck in with the public / private re-shuffle

nwolfe replied:
I just meant that we think this formatting is a good idea

jeffmccune replied:
Oh, heh.  At least we don't have hash rockets to argue about, but I guess let bindings are effectively the same thing.

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -15,53 +15,82 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Private
 
+(schema/defn get-system-env :- {schema/Str schema/Str}
```
Since we have this `{schema/Str schema/Str}` in three places now I _might_ consider a schema for it. Just maybe though

jeffmccune replied:
I'm convinced.  Fixed in 79711eb `(maint) Add jruby-schemas/EnvMap`.

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -15,53 +15,82 @@
+  RubyInstanceConfig$CompileMode/OFF)
+
+(def compat-version
+  "The JRuby compatibility version to use"
```
Your docstrings are kicking butt so far.  I think we should mention in these two that they're used for both the server and CLI.

jeffmccune replied:
Fixed up in 9965399 `(maint) Add compile-mode, compat-version defs`  Thanks, great suggestion.

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -110,6 +139,21 @@
         (.put puppet-config dir (fs/absolute-path value))))
     puppet-config))
 
+(schema/defn borrow-with-timeout-fn :- jruby-schemas/JRubyPuppetBorrowResult
```
You didn't do this, but I'm wondering why this is named as `...-fn`.  Does it return a function?

nwolfe replied:
Don't worry about spending any time tracking this down and potentially changing it unless it's worthwhile.  I'm more just curious

jeffmccune replied:
IDK, but looking at the schema it doesn't look to be a function:

```clojure
(def JRubyPuppetBorrowResult
  (schema/pred (some-fn nil?
                        poison-pill?
                        retry-poison-pill?
                        jruby-puppet-instance?)))
```

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -110,6 +139,21 @@
+
+(schema/defn ^:always-validate
+  create-pool-from-config :- jruby-schemas/PoolState
+  "Create a new PoolData based on the config input."
```
You didn't do this, but looks like we forgot to update the docstring when we renamed `PoolData`.  I might update it, but no big deal

jeffmccune replied:
Fixed up in 54b7bb7 `(maint) Mark j_p_internal functions public / private`

nwolfe commented on src/clj/puppetlabs/services/jruby/jruby_puppet_internal.clj:
```diff
@@ -250,3 +289,18 @@
+(schema/defn ^:always-validate new-main :- jruby-schemas/JRubyMain
+  "Return a new JRuby Main instance which should only be used for CLI purposes,
+  e.g. for the ruby, gem, and irb subcommands.  Internal core services should
+  use `create-scripting-container` instead of `new-jruby-main`."
```
Looks like this is trying to reference itself, which is named `new-main`.

jeffmccune replied:
Fixed up in 321237a `(SERVER-297) Consolidate JRuby environment handling`

nwolfe commented on acceptance/suites/tests/010-puppetserver-cli/subcommand/SERVER-297_common_behavior.rb:
```diff
@@ -0,0 +1,33 @@
+  assert_match(/\bGEM_HOME\b/, output, "GEM_HOME missing")
+  assert_match(/\bJARS_REQUIRE\b/, output, "JARS_REQUIRE missing")
+  assert_match(/\bJARS_NO_REQUIRE\b/, output, "JARS_NO_REQUIRE missing")
+end
```
:v: nice looking test

jeffmccune replied:
Thanks!

nwolfe commented on acceptance/suites/tests/010-puppetserver-cli/subcommand/SERVER-297_common_behavior.rb:
```diff
@@ -0,0 +1,33 @@
+# the need to fix this once Beaker has been modified to make the paths to these
+# commands available.
+#
+cli = "puppetserver"
```
Do we need this comment and `cli` variable anymore?  Let's just inline it below and delete this cruft if so

jeffmccune replied:
Fixed up in b14b711 `(SERVER-297) Add acceptance tests for env behavior`

nwolfe commented on acceptance/suites/tests/010-puppetserver-cli/subcommand/SERVER-297_common_behavior.rb:
```diff
@@ -0,0 +1,33 @@
+require 'test/unit/assertions'
+require 'json'
```
Also doesn't look like this is used?

nwolfe replied:
I'm stupid, it's right there

jeffmccune replied:
It's used for `JSON.parse`

nwolfe commented:
:+1: other than minor nitpick comments

jeffmccune commented:
@nwolfe I've addressed all of your comments.  Thanks for the eye to detail.
